File: bonus/Microsoft/event2/task3.go
package main

import (
	"fmt"
	"math"
)

func main() {
	fmt.Println(Solution([]int{4, 2, 3, 7}, 2, 2))
	fmt.Println(Solution([]int{10, 3, 4, 7}, 2, 3))
	fmt.Println(Solution([]int{4, 2, 5, 4, 3, 5, 1, 4, 2, 7}, 3, 2))
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func dfs(A []int, index int, X int, Y int, memo map[pair]int) int {
	if X < 0 {
		return 0
	}
	if index > len(A)-1 {
		// we can't finish the rehabilitation before X < 0
		return math.MaxInt32
	}
	// check if we have already calculated the cost for this index
	if val, ok := memo[pair{index, X}]; ok {
		fmt.Println("memo", index, val)
		return val
	}
	cost := math.MaxInt32
	cost = min(cost, dfs(A, index+Y, X-1, Y, memo)+A[index])
	memo[pair{index, X}] = cost
	return cost

}

type pair struct {
	x         int
	remaining int
}

func Solution(A []int, X int, Y int) int {
	// write your code in Go (Go 1.4)
	total := math.MaxInt32

	// create a memo
	memo := map[pair]int{}
	for i := 0; i < len(A); i++ {
		cost := dfs(A, i, X-1, Y, memo)
		total = min(total, cost)
	}

	return total
}

File: coding_interviews/Microsoft/event1/task3.py
# you can write to stdout for debugging purposes, e.g.
# print("this is a debug message")


def solution(A, B, S):
    # write your code in Python (Python 3.6)

    def dfs(choosed, slots, i):
        if i > len(A) - 1:
            return True
        # we could choose A[i] or B[i]
        if slots[choosed] == True:
            return False
        slots[choosed] = True

        # check index out of range
        if i > len(A) - 2:
            return True

        return dfs(A[i + 1], slots, i + 1) or dfs(B[i + 1], slots, i + 1)

    slots = {}
    for i in range(S + 1):
        slots[i] = False

    return dfs(B[0], slots, 0) or dfs(A[0], slots, 0)


print(solution([1, 1, 3], [2, 2, 1], 3))
print(solution([3, 2, 3, 1], [1, 3, 1, 2], 3))
print(solution([2, 5, 6, 5], [5, 4, 2, 2], 8))
print(solution([1, 2, 1, 6, 8, 7, 8], [2, 3, 4, 7, 7, 8, 7], 10))
